src/Orleans/Runtime/CallbackData.cs

```diff
@@ -13,30 +15,51 @@ internal interface ITimebound
         /// This method is called by the timer when the time out is reached.
         /// </summary>
         void OnTimeout();
-        TimeSpan RequestedTimeout();
+
+        DateTime DueTime { get; }
```

Comment:
Yes, that's the alternative approach that requires using of the weak references which incurs so much overhead (500 ms per 2 million callbacks) so I'm not convinced that it can be faster than current implementation. 

Agreed about making due time immutable.